src/raftstore/store/fsm/batch.rs

```diff
@@ -0,0 +1,583 @@
+
+    /// This function is called at the end of every round.
+    fn end(&mut self);
+}
+
+/// Internal poller that fetches batch and call handler hooks for readiness.
+struct Poller<N: Fsm, C: Fsm, Handler> {
+    router: Router<N, C, NormalScheduler<N, C>, ControlScheduler<N, C>>,
+    fsm_receiver: channel::Receiver<FsmTypes<N, C>>,
+    handler: Handler,
+    max_batch_size: usize,
+}
+
+impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
+    fn fetch_batch(&self, batch: &mut Batch<N, C>, max_size: usize) {
+        let mut pushed = match self.fsm_receiver.recv() {
+            Ok(fsm) => batch.push(fsm),
+            Err(_) => return,
+        };
+        let mut tried = 1;
+        while pushed {
+            if tried < max_size {
+                let fsm = match self.fsm_receiver.try_recv() {
+                    Ok(fsm) => fsm,
+                    Err(TryRecvError::Empty) => return,
+                    Err(TryRecvError::Disconnected) => unreachable!(),
+                };
+                pushed = batch.push(fsm);
+                tried += 1;
+            } else {
+                return;
+            }
+        }
+        batch.clear();
+    }
+
+    // Poll for readiness and forward to handler. Remove stale peer if necessary.
+    fn poll(&mut self) {
+        let mut batch_size = self.max_batch_size;
+        let mut batch = Batch::with_capacity(batch_size);
+        let mut exhausted_fsms = Vec::with_capacity(batch_size);
+        self.fetch_batch(&mut batch, batch_size);
+        while !batch.is_empty() {
+            self.handler.begin();
+            if batch.control.is_some() {
+                let pos = self.handler.handle_control(batch.control.as_mut().unwrap());
+                if batch.control.as_ref().unwrap().is_stopped() {
+                    batch.remove_control(&self.router.control_box);
+                } else if let Some(pos) = pos {
+                    batch.release_control(&self.router.control_box, pos);
+                }
+            }
+            if !batch.normals.is_empty() {
+                for (i, p) in batch.normals.iter_mut().enumerate() {
+                    let pos = self.handler.handle_normal(p);
+                    if p.is_stopped() {
+                        exhausted_fsms.push((i, None));
+                    } else if pos.is_some() {
+                        exhausted_fsms.push((i, pos));
+                    }
+                }
+            }
+            self.handler.end();
+            while let Some((r, mark)) = exhausted_fsms.pop() {
+                if let Some(m) = mark {
+                    batch.release(r, m);
+                } else {
+                    batch.remove(r);
+                }
+            }
+            if batch.is_empty() {
+                batch_size = cmp::min(batch_size + 1, self.max_batch_size);
+                self.fetch_batch(&mut batch, batch_size);
+            } else {
+                batch_size = cmp::max(batch_size - 1, 1);
```

Comment:
Why `-1`?

Reply:
If the batch is not finished, then there can be write skew, use a smaller batch to avoid unbalanced load.